druid/src/window.rs

```diff
@@ -288,6 +293,8 @@ impl<T: Data> Window<T> {
         if ctx.base_state.request_anim {
             self.last_anim = Some(now);
         }
+
+        self.post_event_processing(&mut base_state, queue, data, env, process_commands);
```

Comment:
Is there a specific reason we think this is necessary? do_anim_frame is part of paint, and paint calls this already.

Reply:
- It's necessary because a widget might call `submit_command` or `request_timer` in response to `AnimFrame`.
- `do_anim_frame` is currently part of `lifecycle` not `paint` and so even if `paint` called `post_event_processing` that wouldn't be guaranteed for other `AnimFrame` calls to `lifecycle`.
- `paint` doesn't call `post_event_processing`.
- `AnimFrame` does not guarantee `paint` in the future as discussed [on zulip](https://xi.zulipchat.com/#narrow/stream/147926-druid/topic/purpose.20of.20request_anim).

Reply:
in order:

- to my mind, `AnimFrame` should only be used to update state in preparation for an imminent paint call. This is a weird case;  `AnimFrame` is one of the bits of event flow that is really a poor fit to the current architecture, and when (like.... a year ago? 😬) I was playing around with a higher-level animation API, one of the changes was to just make a `prepare_for_animation` method on `Widget` that replaced `AnimFrame`. `submit_command` and `request_timer` should be discouraged here, and it's annoying that this isn't enforced by the type system.

- `do_anim_frame` sort of is part of paint; `LifeCycle::AnimFrame` should only be sent from `Window::do_paint`, if it happens anywhere else that's a mistake.

- this checks out, I must've misread again. That said, it looks like we  call `post_event_processing` in Window::lifecycle`, which is what calls `do_anim_frame`?

- I think this is false? it's not really that anim guarantees `paint`, it's that `AnimFrame` is only called from the painting code path.

Reply:
* Yes we do call `post_event_processing` in `Window::lifecycle` right now, which means it also applies to `do_anim_frame`. However due to borrow checking limitations, in order to keep that functionality, I had to add it to `do_anim_frame` because the `Window::lifecycle` call moved into an `if` block.

* Yeah right now `AnimFrame` is sent just before `paint`. However based on that zulip discussion this may change. @raphlinus said *It's up to the handler of that lifecycle event to request painting* - from which I derive that if the handler doesn't request painting, the painting might not happen at all. (This is for future behavior, today that's not true.)

More generally, moving `AnimFrame` out of `lifecycle` makes sense if it's supposed to have different limitations. However as long as it's part of `lifecycle` I think it should behave like a `lifecycle` event. That includes a working `request_timer` and `submit_command`.

Reply:
On the second point: I think there's a confusion about raph's comments. `AnimFrame` _does_ mean you're going to paint; there's no short-circuiting. If you see it, it means a call to paint is next. To continue the animation, you *do* need to request another frame.

I'll take a look at the code you're talking about, I might've missed that.